chemclipse/plugins/org.eclipse.chemclipse.msd.model/src/org/eclipse/chemclipse/msd/model/matrix/ExtractedMatrix.java

```diff
@@ -0,0 +1,189 @@
+					}
+				}
+			} catch (Exception e) {
+				e.printStackTrace();
+			}
+		}
+	}
+	
+	private Boolean checkHighRes( int limit ) {
+		boolean moreIonsThanFullMz = false;
+		Iterator<IScanMSD> iterator = scans.iterator();
+		IIonBounds bounds;
+		double rangeAbs;
+		IScanMSD currentScan;
+		
+		while(moreIonsThanFullMz == false) {
+			currentScan = iterator.next();
+			bounds = currentScan.getIonBounds();
+			rangeAbs = bounds.getHighestIon().getIon() - bounds.getLowestIon().getIon();
+			if(rangeAbs + limit < currentScan.getIons().size()) {
+				moreIonsThanFullMz = true;
+			}
+			if(!iterator.hasNext()) {
+				break;
+			}
+		}
+		
+		return moreIonsThanFullMz;
+	}
+	
+	private List<IScanMSD> extractScans() {
+
+		List<IScanMSD> scans;
+		int startRT;
+		int stopRT;
+		startRT = selection.getStartRetentionTime();
+		stopRT = selection.getStopRetentionTime();
+		scans = selection.getChromatogram() //
+				.getScans() //
+				.stream() //
+				.filter(s -> s instanceof IScanMSD) //
+				.map(IScanMSD.class::cast) //
+				.filter(s -> s.getRetentionTime() > startRT) //
+				.filter(s -> s.getRetentionTime() < stopRT) //
+				.collect(Collectors.toList()); //
+		return (scans);
+	}
+	
+	private int getMinMz() {
+
+		double min = scans.stream() //
+				.flatMap(scan -> scan.getIons().stream()) //
+				.min(Comparator.comparing(IIon::getIon)) //
+				.get() //
+				.getIon();
+		int minMz = (int)Math.round(min);
+		return (minMz);
+	}
+	
+	private int getMaxMz() {
+
+		double max = scans.stream() //
+				.flatMap(scan -> scan.getIons().stream()) //
+				.max(Comparator.comparing(IIon::getIon)) //
+				.get() //
+				.getIon();
+		int maxMz = (int)Math.round(max);
+		return (maxMz);
+	}
+
+	@Override
+	public float[] getMatrix() {
+		if(highRes == true) {
+			return null;			
+		}
+		return this.signal;
+	}
+
+	@Override
+	public int getNumberOfScans() {
+		if(highRes == true) {
+			return 0;			
+		}
+		return this.numberOfScans;
+	}
+
+	@Override
+	public int getNumberOfIons() {
+		if(highRes == true) {
+			return 0;			
+		}
+		return this.numberOfIons;
+	}
+
+	@Override
+	public void updateSignal(float[] signal, int numberOfScans, int nubmerOfIons ) {
+		
+		IScanMSD currentScan;
+		IIon currentIon;
+		
+		if(highRes != true) {
+			try {
+				for(int i = 1; i <= numberOfScans; i++) {
+					currentScan = (IScanMSD) selection.getChromatogram().getScan(i);
+					currentScan.removeAllIons();
+					for(int j = startIon; j < stopIon; j++) {
+						if(signal[(j - startIon) * numberOfScans + (i-1)] != 0.0) {
+							currentIon = new Ion(j, signal[(j - startIon) * numberOfScans + (i - 1)] ); 
+							currentScan.addIon(currentIon);
+						}		
+					}
+				}
+			} catch (Exception e) {
+				e.printStackTrace();
+			}			
+		}
+	}
+
+	@Override
+	public int[] getScanNumbers() {
+		if(highRes == true) {
+			return null;			
+		}
+		int[] scanNumbers = scans.stream().mapToInt(scan -> scan.getScanNumber()).toArray();
```

Comment:
You could simply return an int[][] here to include scan number together with scan retentiontimes.